### PR TITLE
Extmarks api: allow for gravity

### PR DESCRIPTION
**Pull request by chentoast**
Allows for extmark gravity to be set.

An example: suppose we have the following text, with an extmark placed where the pipe is:

`asdf|asdf`

normally, if we type (gravity = "right"), the extmark will move with the cursor:

`asdfhello|asdf`

but with gravity = "left", the extmark will stay put:

`asdf|helloasdf`.

To decide: do we want to allow gravity for the end_col and end_row of extmarks as well? Would be useful to know of potential use cases where this would be needed.

### CODE REVIEW

**Review by vigoux — APPROVED** [review]

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -1522,6 +1526,19 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
         goto error;
       }
       priority = (DecorPriority)v->data.integer;
+    } else if (strequal("gravity", k.data)) {
+      if (v->type != kObjectTypeString) {
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1530]: better to make it a bool, there will only be two values.

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -1400,6 +1400,9 @@ Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id,
 ///                   callbacks. The mark will only be used for the current
 ///                   redraw cycle, and not be permantently stored in the
 ///                   buffer.
+///               - gravity : the direction the extmark will be shifted in
+///                   when new text is inserted. Must be either 'left' or
+///                   'right'.
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1405]: there should also be `end_gravity` for completeness (last arg to `marktree_put_pair`)

**Comment by chentau** [comment]
> better to make it a bool, there will only be two values.

There was a helpful discussion about this in #12249. In the end, I felt that two options seemed the best to me:

```
{gravity = "right"}, {gravity = "left"}
```
or
```
{right_gravity = true}, {right_gravity = false}
```

Either option is fine with me. Once we decide on which one to go with, I'll add the corresponding option for end_gravity.

**Comment by bfredl** [comment]
`right_gravity=true` is better.

**Comment by runiq** [comment]
What about RTL (if it's ever going to be supported with extmarks, that is)? Would it make sense for RTL text to have `right_gravity=false` by default?

**Comment by runiq** [comment]
Friendly reminder: https://github.com/neovim/neovim/pull/12742 is still open, if you find anything useful in there, feel free to steal :)

**Comment by bfredl** [comment]
> What about RTL (if it's ever going to be supported with extmarks, that is)? Would it make sense for RTL text to have right_gravity=false by default?

RTL is a display feature (that UI:s are gonna handle in differerent ways), it better not affect editing semantics in the core layer.

File: src/nvim/api/buffer.c [hunk 3]
```diff
@@ -1546,6 +1570,14 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
     col2 = 0;
   }
 
+  // Only error out if they try to set end_right_gravity to true without
+  // setting end_col or end_line
+  if (line2 == 0 && col2 == 0 && end_right_gravity) {
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1575]: hmm I think it is `line2 == -1`

File: src/nvim/api/buffer.c [hunk 4]
```diff
@@ -1546,6 +1570,14 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
...
+    api_set_error(err, kErrorTypeValidation,
+                  "cannot set end_right_gravity to true "
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1577]: "cannot set end_right_gravity" is enough, neither value makes sense if end doesn't exist.

**Comment by chentau** [comment]
@bfredl whenever you have the time, do you mind reviewing this? I think its pretty much ready for merge.

**Comment by bfredl** [comment]
I agree, merged :)

**Pull request merged**